pkg/scripts/install.go

```diff
@@ -13,177 +15,230 @@ import (
 	"github.com/openshift/source-to-image/pkg/util"
 )
 
-// Installer interface is responsible for installing scripts needed to run the build
+// Installer interface is responsible for installing scripts needed to run the
+// build.
 type Installer interface {
 	InstallRequired(scripts []string, dstDir string) ([]api.InstallResult, error)
 	InstallOptional(scripts []string, dstDir string) []api.InstallResult
 }
 
-// NewInstaller returns a new instance of the default Installer implementation
-func NewInstaller(image string, scriptsURL string, docker docker.Docker, auth dockerClient.AuthConfiguration) Installer {
-	return &installer{
-		image:      image,
-		scriptsURL: scriptsURL,
-		docker:     docker,
-		downloader: NewDownloader(),
-		pullAuth:   auth,
-		fs:         util.NewFileSystem(),
-	}
+// ScriptHandler provides an interface for various scripts source handlers.
+type ScriptHandler interface {
+	Get(script string) *api.InstallResult
+	Install(*api.InstallResult) error
+	SetDestinationDir(string)
 }
 
-type installer struct {
-	image      string
-	scriptsURL string
-	docker     docker.Docker
-	pullAuth   dockerClient.AuthConfiguration
-	downloader Downloader
-	fs         util.FileSystem
+// UrlScriptHandler handles script download using URL.
+type UrlScriptHandler struct {
+	URL            string
+	DestinationDir string
+	download       Downloader
+	fs             util.FileSystem
 }
 
-// locationsOrder defines locations in which scripts are searched for in the following order:
-// - script found at the --scripts-url URL
-// - script found in the application source .sti/bin directory
-// - script found at the default image URL
-var locationsOrder = []string{api.UserScripts, api.SourceScripts, api.DefaultScripts}
+// SetDestinationDir sets the destination where the scripts should be
+// downloaded.
+func (s *UrlScriptHandler) SetDestinationDir(baseDir string) {
+	s.DestinationDir = filepath.Join(baseDir, api.UploadScripts)
+}
 
-// InstallRequired downloads and installs required scripts into dstDir, the result is a
-// map of scripts with detailed information about each of the scripts install process
-// with error if installing some of them failed
-func (i *installer) InstallRequired(scripts []string, dstDir string) (results []api.InstallResult, err error) {
-	results = i.run(scripts, dstDir)
-	failedScripts := []string{}
-	for _, r := range results {
-		if !r.Installed && r.Error != nil {
-			failedScripts = append(failedScripts, r.Script)
-		}
+// Get parses the provided URL and the script name.
+func (s *UrlScriptHandler) Get(script string) *api.InstallResult {
```

Comment:
why does Get take an InstallResult?  and why does Install take an InstallResult?  i'd expect it to return one.